### ISSUE TITLE
item in has_many collection not being destroyed

### ISSUE THREAD

**Issue by kevinansfield** [apotonick/disposable#3]
I have what I believe is a pretty direct translation of the collection item removal from the gemgem code but for some reason the record from the collection is never actually destroyed.

I've put all the related code in this gist and added comments to the item deletion test showing the values I'm seeing in the reform/disposable instance at each stage: https://gist.github.com/kevinansfield/5ac5c612921418ea874a#file-crud_test-rb-L72

Do you have any idea what's causing the deletion to fail? Thanks!

**Comment by kevinansfield** [comment]
I've verified that `#destroy` is actually being called on the model instances by so I'm not sure this is actually a disposable issue. Although I'm now left even more clueless!

**Comment by apotonick** [comment]
Did you `include D:T:Collection::Semantics` into the twin/form? This is still beta and I am thinking of including that automatically. This will actually do call `destroy` on the items.

**Comment by kevinansfield** [comment]
Thanks. If I include that then the twin gets moved into the `destroyed` array but the test still fails as the actual model record isn't destroyed even though `#destroy` is being called on it.

I'm beginning to wonder whether the test being wrapped in a transaction is causing an issue somewhere. Once I've worked out how to get around an `undefined method 'new_record?'` issue with AA/formtastic then I'll be able to test it manually to see if that sheds any new light.

**Comment by kevinansfield** [comment]
I've updated the gist and switched back to using `moves.delete` as per the gemgem source. The record is actually being destroyed however as you can see from the debug output the next record in the collection takes it's place (and isn't updated) then the third is updated and finally a new record added at the end: https://gist.github.com/kevinansfield/5ac5c612921418ea874a#file-crud_test-rb-L202

It's almost as if the passed-in IDs are not being used.

**Comment by apotonick** [comment]
Disposable's collection semantic is very simple. It works as follows.

```ruby
comment.users = [1,2,3]
comment.users.delete(1)
comment.users #=> [2,3]

comment.save #=> model.users = comment.users
```

Does your behaviour make more sense, now?

**Comment by kevinansfield** [comment]
I understand that but that doesn't seem to be what's happening. Whilst debugging the item is definitely removed from the collection in the `skip_move?` and it's still removed when going through the populator step in representable, but for some reason once it gets to the mapping stage a duplicate of the first item of the array is added but with the original values.

The models associated with each item in the collection also seem to become out of sync as soon as one of the items is deleted, is that normal?

```ruby
def skip_move?(fragment, options)
  byebug if fragment['_destroy']
  return true if fragment['_destroy'] == '1' && moves.delete(moves.find { |m| m.id.to_s == fragment['id'] })
  return true if fragment['name'].blank?
end
```

---

```
➜  competio-app git:(enforce-judging-completion) ✗ ruby -I test test/concepts/competition/crud_test.rb
@@@@@!!!! {:instance=>#<Reform::Form::Populator::IfEmpty:0x007fbe3756bd30 @user_proc=Move (call 'Move.connection' to establish a connection), @value=#<Uber::Options::Value:0x007fbe3756bd08 @dynamic=false, @value=Move (call 'Move.connection' to establish a connection), @proc=false, @callable=false, @method=false>>, :setter=>nil, :skip_parse=>#<Reform::Form::Validate::Skip::AllBlank:0x007fbe3756bc40>}
@@@@@!!!! {:instance=>#<Reform::Form::Populator::IfEmpty:0x007fbe3756bd30 @user_proc=Move (call 'Move.connection' to establish a connection), @value=#<Uber::Options::Value:0x007fbe3756bd08 @dynamic=false, @value=Move (call 'Move.connection' to establish a connection), @proc=false, @callable=false, @method=false>>, :setter=>nil, :skip_parse=>:skip_move?, :writeable=>true}

# Running tests with run options --seed 16710:

.....
[97, 106] in /Users/kevinansfield/code/rails/competio-app/app/concepts/competition/crud.rb
    97: 
    98:       private
    99: 
   100:       def skip_move?(fragment, options)
   101:         byebug if fragment['_destroy']
=> 102:         return true if fragment['_destroy'] == '1' && moves.delete(moves.find { |m| m.id.to_s == fragment['id'] })
   103:         return true if fragment['name'].blank?
   104:       end
   105:     end
   106: 
(byebug) moves.size
3
(byebug) next

[82, 91] in /Users/kevinansfield/.rvm/gems/ruby-2.2.2/gems/representable-2.2.2/lib/representable/deserializer.rb
   82:       def call(fragment)
   83:         collection = [] # this can be replaced, e.g. AR::Collection or whatever.
   84: 
   85:         fragment.each_with_index do |item_fragment, i|
   86:           # add more per-item options here!
=> 87:           next if @binding.evaluate_option(:skip_parse, item_fragment) # TODO: pass in index!
   88: 
   89:           collection << deserialize!(item_fragment, i) # FIXME: what if obj nil?
   90:         end
   91: 
(byebug) next

[84, 93] in /Users/kevinansfield/.rvm/gems/ruby-2.2.2/gems/representable-2.2.2/lib/representable/deserializer.rb
   84: 
   85:         fragment.each_with_index do |item_fragment, i|
   86:           # add more per-item options here!
   87:           next if @binding.evaluate_option(:skip_parse, item_fragment) # TODO: pass in index!
   88: 
=> 89:           collection << deserialize!(item_fragment, i) # FIXME: what if obj nil?
   90:         end
   91: 
   92:         collection # with parse_strategy: :sync, this is ignored.
   93:       end
(byebug) next

[82, 91] in /Users/kevinansfield/.rvm/gems/ruby-2.2.2/gems/representable-2.2.2/lib/representable/deserializer.rb
   82:       def call(fragment)
   83:         collection = [] # this can be replaced, e.g. AR::Collection or whatever.
   84: 
   85:         fragment.each_with_index do |item_fragment, i|
   86:           # add more per-item options here!
=> 87:           next if @binding.evaluate_option(:skip_parse, item_fragment) # TODO: pass in index!
   88: 
   89:           collection << deserialize!(item_fragment, i) # FIXME: what if obj nil?
   90:         end
   91: 
(byebug) next

[84, 93] in /Users/kevinansfield/.rvm/gems/ruby-2.2.2/gems/representable-2.2.2/lib/representable/deserializer.rb
   84: 
   85:         fragment.each_with_index do |item_fragment, i|
   86:           # add more per-item options here!
   87:           next if @binding.evaluate_option(:skip_parse, item_fragment) # TODO: pass in index!
   88: 
=> 89:           collection << deserialize!(item_fragment, i) # FIXME: what if obj nil?
   90:         end
   91: 
   92:         collection # with parse_strategy: :sync, this is ignored.
   93:       end
(byebug) next

[87, 96] in /Users/kevinansfield/.rvm/gems/ruby-2.2.2/gems/representable-2.2.2/lib/representable/deserializer.rb
   87:           next if @binding.evaluate_option(:skip_parse, item_fragment) # TODO: pass in index!
   88: 
   89:           collection << deserialize!(item_fragment, i) # FIXME: what if obj nil?
   90:         end
   91: 
=> 92:         collection # with parse_strategy: :sync, this is ignored.
   93:       end
   94: 
   95:     private
   96:       def deserialize!(*args)
(byebug) pp collection
[#<#<Class:0x007fbe39246568>:0x007fbe3d0d1af8
  @_changes={"name"=>true, "position"=>true, "id"=>true},
  @errors=
   #<Reform::Contract::Errors:0x007fbe33c73ac8
    @base=#<#<Class:0x007fbe39246568>:0x007fbe3d0d1af8 ...>,
    @messages={}>,
  @fields={"name"=>"Middle Changed", "position"=>2, "id"=>"908005751"},
  @model=
   #<Move:0x007fbe3d0d29d0
    id: 908005749,
    name: "Last",
    position: 3,
    competition_id: 802290395,
    created_at: Sun, 14 Jun 2015 08:37:27 UTC +00:00,
    updated_at: Sun, 14 Jun 2015 08:37:27 UTC +00:00>,
  @validation_context=nil>,
 #<#<Class:0x007fbe39246568>:0x007fbe3d1600a0
  @_changes={"name"=>true, "position"=>true, "id"=>true},
  @errors=
   #<Reform::Contract::Errors:0x007fbe3d2d0f70
    @base=#<#<Class:0x007fbe39246568>:0x007fbe3d1600a0 ...>,
    @messages={}>,
  @fields={"name"=>"Last Changed", "position"=>3, "id"=>"908005749"},
  @model=
   #<Move:0x007fbe3d150d30
    id: nil,
    name: nil,
    position: nil,
    competition_id: nil,
    created_at: nil,
    updated_at: nil>,
  @validation_context=nil>]
(byebug) next

[19, 28] in /Users/kevinansfield/.rvm/gems/ruby-2.2.2/gems/representable-2.2.2/lib/representable/populator.rb
   19:       else
   20:         # DISCUSS: should we return a Skip object instead of this block trick? (same in Binding#serialize?)
   21:         value = deserialize(fragment) { return } # stop here if skip_parse?
   22:       end
   23: 
=> 24:       value = @binding.parse_filter(value, doc)
   25:         # parse_filter
   26:         # set
   27:       @binding.set(value)
   28:     end
(byebug) next

[22, 31] in /Users/kevinansfield/.rvm/gems/ruby-2.2.2/gems/representable-2.2.2/lib/representable/populator.rb
   22:       end
   23: 
   24:       value = @binding.parse_filter(value, doc)
   25:         # parse_filter
   26:         # set
=> 27:       @binding.set(value)
   28:     end
   29: 
   30:   private
   31:     def deserialize(fragment)
(byebug) pp value
[#<#<Class:0x007fbe39246568>:0x007fbe3d0d1af8
  @_changes={"name"=>true, "position"=>true, "id"=>true},
  @errors=
   #<Reform::Contract::Errors:0x007fbe33c73ac8
    @base=#<#<Class:0x007fbe39246568>:0x007fbe3d0d1af8 ...>,
    @messages={}>,
  @fields={"name"=>"Middle Changed", "position"=>2, "id"=>"908005751"},
  @model=
   #<Move:0x007fbe3d0d29d0
    id: 908005749,
    name: "Last",
    position: 3,
    competition_id: 802290395,
    created_at: Sun, 14 Jun 2015 08:37:27 UTC +00:00,
    updated_at: Sun, 14 Jun 2015 08:37:27 UTC +00:00>,
  @validation_context=nil>,
 #<#<Class:0x007fbe39246568>:0x007fbe3d1600a0
  @_changes={"name"=>true, "position"=>true, "id"=>true},
  @errors=
   #<Reform::Contract::Errors:0x007fbe3d2d0f70
    @base=#<#<Class:0x007fbe39246568>:0x007fbe3d1600a0 ...>,
    @messages={}>,
  @fields={"name"=>"Last Changed", "position"=>3, "id"=>"908005749"},
  @model=
   #<Move:0x007fbe3d150d30
    id: nil,
    name: nil,
    position: nil,
    competition_id: nil,
    created_at: nil,
    updated_at: nil>,
  @validation_context=nil>]
(byebug) next

[15, 24] in /Users/kevinansfield/.rvm/gems/ruby-2.2.2/gems/representable-2.2.2/lib/representable/mapper.rb
   15: 
   16:       def deserialize(represented, doc, options, private_options)
   17:         bindings(represented, options).each do |bin|
   18:           deserialize_property(bin, doc, options, private_options)
   19:         end
=> 20:         represented
   21:       end
   22: 
   23:       def serialize(represented, doc, options, private_options)
   24:         bindings(represented, options).each do |bin|
(byebug) pp represented.moves
[#<#<Class:0x007fbe39246568>:0x007fbe3d0d20c0
  @_changes={},
  @fields={"name"=>"Middle", "position"=>2, "id"=>908005751},
  @model=
   #<Move:0x007fbe3d0d2b38
    id: 908005751,
    name: "Middle",
    position: 2,
    competition_id: 802290395,
    created_at: Sun, 14 Jun 2015 08:37:27 UTC +00:00,
    updated_at: Sun, 14 Jun 2015 08:37:27 UTC +00:00>>,
 #<#<Class:0x007fbe39246568>:0x007fbe3d0d1af8
  @_changes={"name"=>true, "position"=>true, "id"=>true},
  @errors=
   #<Reform::Contract::Errors:0x007fbe33c73ac8
    @base=#<#<Class:0x007fbe39246568>:0x007fbe3d0d1af8 ...>,
    @messages={}>,
  @fields={"name"=>"Middle Changed", "position"=>2, "id"=>"908005751"},
  @model=
   #<Move:0x007fbe3d0d29d0
    id: 908005749,
    name: "Last",
    position: 3,
    competition_id: 802290395,
    created_at: Sun, 14 Jun 2015 08:37:27 UTC +00:00,
    updated_at: Sun, 14 Jun 2015 08:37:27 UTC +00:00>,
  @validation_context=nil>,
 #<#<Class:0x007fbe39246568>:0x007fbe3d1600a0
  @_changes={"name"=>true, "position"=>true, "id"=>true},
  @errors=
   #<Reform::Contract::Errors:0x007fbe3d2d0f70
    @base=#<#<Class:0x007fbe39246568>:0x007fbe3d1600a0 ...>,
    @messages={}>,
  @fields={"name"=>"Last Changed", "position"=>3, "id"=>"908005749"},
  @model=
   #<Move:0x007fbe3d150d30
    id: nil,
    name: nil,
    position: nil,
    competition_id: nil,
    created_at: nil,
    updated_at: nil>,
  @validation_context=nil>]
  (byebug) next

[20, 29] in /Users/kevinansfield/.rvm/gems/ruby-2.2.2/bundler/gems/reform-70c02d1743f3/lib/reform/form/validate.rb
   20:   # 2. Deserialize. This is wrong and should be done in 1.
   21:   # 3. Validate the form object graph.
   22:   def validate(params)
   23:     update!(params)
   24: 
=> 25:     super() # run the actual validation on self.
   26: 
   27:   # rescue Representable::DeserializeError
   28:   #   raise DeserializeError.new("[Reform] Deserialize error: You probably called #validate without setting up your nested models. Check https://github.com/apotonick/reform#populating-forms-for-validation on how to use populators.")
   29:   end
(byebug) pp params
{:name=>"Updated Competition",
 :description=>"I am a new description.",
 :start_at=>Wed, 01 Jul 2015 00:00:00 UTC +00:00,
 :end_at=>Thu, 30 Jul 2015 23:59:59 UTC +00:00,
 :example_video=>{"video_file"=>""},
 "moves"=>
  [{"id"=>"908005750",
    "position"=>1,
    "name"=>"First Changed",
    "_destroy"=>"1"},
   {"id"=>"908005751", "position"=>2, "name"=>"Middle Changed"},
   {"id"=>"908005749", "position"=>3, "name"=>"Last Changed"}]}
(byebug) next

[106, 115] in /Users/kevinansfield/.rvm/gems/ruby-2.2.2/bundler/gems/trailblazer-107cea3496dc/lib/trailblazer/operation.rb
   106: 
   107:     def validate(params, model, contract_class=nil)
   108:       @contract = contract_for(contract_class, model)
   109: 
   110:       if @valid = contract.validate(params)
=> 111:         yield contract if block_given?
   112:       else
   113:         raise!(contract)
   114:       end
   115: 
(byebug) pp params
{:name=>"Updated Competition",
 :description=>"I am a new description.",
 :start_at=>Wed, 01 Jul 2015 00:00:00 UTC +00:00,
 :end_at=>Thu, 30 Jul 2015 23:59:59 UTC +00:00,
 :example_video=>{"video_file"=>""},
 "moves"=>
  [{"id"=>"908005750",
    "position"=>1,
    "name"=>"First Changed",
    "_destroy"=>"1"},
   {"id"=>"908005751", "position"=>2, "name"=>"Middle Changed"},
   {"id"=>"908005749", "position"=>3, "name"=>"Last Changed"}]}
(byebug) next

[111, 120] in /Users/kevinansfield/.rvm/gems/ruby-2.2.2/bundler/gems/trailblazer-107cea3496dc/lib/trailblazer/operation.rb
   111:         yield contract if block_given?
   112:       else
   113:         raise!(contract)
   114:       end
   115: 
=> 116:       @valid
   117:     end
   118: 
   119:     def invalid!(result=self)
   120:       @valid = false
(byebug) pp @valid
true
(byebug) next

[63, 72] in /Users/kevinansfield/.rvm/gems/ruby-2.2.2/bundler/gems/trailblazer-107cea3496dc/lib/trailblazer/operation.rb
   63:     def run(*params)
   64:       setup!(*params) # where do we assign/find the model?
   65: 
   66:       process(*params)
   67: 
=> 68:       [valid?, self]
   69:     end
   70: 
   71:     def present(*params)
   72:       setup!(*params)
(byebug) pp self
#<Competition::Update:0x007fbe3d0cb248
 @contract=
  #<#<Class:0x007fbe3756ab10>:0x007fbe3d0c9308
   @_changes=
    {"name"=>true, "description"=>true, "start_at"=>true, "end_at"=>true},
   @_mounters=
    {:pattern=>
      #<CarrierWave::Mount::Mounter:0x007fbe3d0c8188
       @column=:pattern,
       @integrity_error=nil,
       @options={},
       @processing_error=nil,
       @record=#<#<Class:0x007fbe3756ab10>:0x007fbe3d0c9308 ...>,
       @uploader=
        #<PatternUploader:0x007fbe3d0c8138
         @model=#<#<Class:0x007fbe3756ab10>:0x007fbe3d0c9308 ...>,
         @mounted_as=:pattern>,
       @uploader_options={:mount_on=>nil}>},
   @errors=
    #<Reform::Contract::Errors:0x007fbe32eef708
     @base=#<#<Class:0x007fbe3756ab10>:0x007fbe3d0c9308 ...>,
     @messages={}>,
   @fields=
    {"discipline_class_id"=>517429222,
     "judge_id"=>301712148,
     "name"=>"Updated Competition",
     "description"=>"I am a new description.",
     "start_at"=>Wed, 01 Jul 2015 00:00:00 UTC +00:00,
     "end_at"=>Thu, 30 Jul 2015 23:59:59 UTC +00:00,
     "example_video"=>
      #<#<Class:0x007fbe391e7900>:0x007fbe3c65ea30
       @_changes={"video_file"=>true},
       @errors=
        #<Reform::Contract::Errors:0x007fbe36d5a998
         @base=#<#<Class:0x007fbe391e7900>:0x007fbe3c65ea30 ...>,
         @messages={}>,
       @fields={"video_file"=>""},
       @model=
        #<Video:0x007fbe3c65ec10
         id: 966191006,
         meta_info: {},
         created_at: Sun, 14 Jun 2015 08:37:27 UTC +00:00,
         updated_at: Sun, 14 Jun 2015 08:37:27 UTC +00:00,
         videoable_id: 802290395,
         videoable_type: "Competition",
         video_file_id: nil,
         video_file_filename: nil,
         video_file_content_type: nil>,
       @validation_context=nil>,
     "moves"=>
      [#<#<Class:0x007fbe39246568>:0x007fbe3d0d20c0
        @_changes={},
        @errors=
         #<Reform::Contract::Errors:0x007fbe39974470
          @base=#<#<Class:0x007fbe39246568>:0x007fbe3d0d20c0 ...>,
          @messages={}>,
        @fields={"name"=>"Middle", "position"=>2, "id"=>908005751},
        @model=
         #<Move:0x007fbe3d0d2b38
          id: 908005751,
          name: "Middle",
          position: 2,
          competition_id: 802290395,
          created_at: Sun, 14 Jun 2015 08:37:27 UTC +00:00,
          updated_at: Sun, 14 Jun 2015 08:37:27 UTC +00:00>,
        @validation_context=nil>,
       #<#<Class:0x007fbe39246568>:0x007fbe3d0d1af8
        @_changes={"name"=>true, "position"=>true, "id"=>true},
        @errors=
         #<Reform::Contract::Errors:0x007fbe33c73ac8
          @base=#<#<Class:0x007fbe39246568>:0x007fbe3d0d1af8 ...>,
          @messages={}>,
        @fields={"name"=>"Middle Changed", "position"=>2, "id"=>"908005751"},
        @model=
         #<Move:0x007fbe3d0d29d0
          id: 908005749,
          name: "Middle Changed",
          position: 2,
          competition_id: 802290395,
          created_at: Sun, 14 Jun 2015 08:37:27 UTC +00:00,
          updated_at: Sun, 14 Jun 2015 08:44:44 UTC +00:00>,
        @validation_context=nil>,
       #<#<Class:0x007fbe39246568>:0x007fbe3d1600a0
        @_changes={"name"=>true, "position"=>true, "id"=>true},
        @errors=
         #<Reform::Contract::Errors:0x007fbe3d2d0f70
          @base=#<#<Class:0x007fbe39246568>:0x007fbe3d1600a0 ...>,
          @messages={}>,
        @fields={"name"=>"Last Changed", "position"=>3, "id"=>"908005749"},
        @model=
         #<Move:0x007fbe3d150d30
          id: 908005752,
          name: "Last Changed",
          position: 3,
          competition_id: 802290395,
          created_at: Sun, 14 Jun 2015 08:44:44 UTC +00:00,
          updated_at: Sun, 14 Jun 2015 08:44:44 UTC +00:00>,
        @validation_context=nil>]},
   @model=
    #<Competition:0x007fbe3d0c94c0
     id: 802290395,
     discipline_class_id: 517429222,
     judge_id: 301712148,
     name: "Updated Competition",
     description: "I am a new description.",
     pattern: nil,
     start_at: Wed, 01 Jul 2015 00:00:00 UTC +00:00,
     end_at: Thu, 30 Jul 2015 23:59:59 UTC +00:00,
     created_at: Sun, 14 Jun 2015 08:37:27 UTC +00:00,
     updated_at: Sun, 14 Jun 2015 08:44:44 UTC +00:00>,
   @validation_context=nil>,
 @model=
  #<Competition:0x007fbe3d0c94c0
   id: 802290395,
   discipline_class_id: 517429222,
   judge_id: 301712148,
   name: "Updated Competition",
   description: "I am a new description.",
   pattern: nil,
   start_at: Wed, 01 Jul 2015 00:00:00 UTC +00:00,
   end_at: Thu, 30 Jul 2015 23:59:59 UTC +00:00,
   created_at: Sun, 14 Jun 2015 08:37:27 UTC +00:00,
   updated_at: Sun, 14 Jun 2015 08:44:44 UTC +00:00>,
 @raise_on_invalid=false,
 @valid=true>
  ```

**Comment by kevinansfield** [comment]
Sorry if that's a bit hard to follow. These are the final states during the `process` stage:

```ruby
def process(params)
  validate(params[:competition]) do |f|
    pp params[:competition]
    # {:name=>"Updated Competition",
    #  :description=>"I am a new description.",
    #  :start_at=>Wed, 01 Jul 2015 00:00:00 UTC +00:00,
    #  :end_at=>Thu, 30 Jul 2015 23:59:59 UTC +00:00,
    #  :example_video=>{"video_file"=>""},
    #  "moves"=>
    #   [{"id"=>"908005747",
    #     "position"=>1,
    #     "name"=>"First Changed",
    #     "_destroy"=>"1"},
    #    {"id"=>"908005748", "position"=>2, "name"=>"Middle Changed"},
    #    {"id"=>"908005746", "position"=>3, "name"=>"Last Changed"}]}
    
    pp f
    # #<#<Class:0x007fa73fbc92e0>:0x007fa73d3d8150
    #  @_changes=
    #   {"name"=>true, "description"=>true, "start_at"=>true, "end_at"=>true},
    #  @_mounters=
    #   {:pattern=>
    #     #<CarrierWave::Mount::Mounter:0x007fa743496f00
    #      @column=:pattern,
    #      @integrity_error=nil,
    #      @options={},
    #      @processing_error=nil,
    #      @record=#<#<Class:0x007fa73fbc92e0>:0x007fa73d3d8150 ...>,
    #      @uploader=
    #       #<PatternUploader:0x007fa743496eb0
    #        @model=#<#<Class:0x007fa73fbc92e0>:0x007fa73d3d8150 ...>,
    #        @mounted_as=:pattern>,
    #      @uploader_options={:mount_on=>nil}>},
    #  @errors=
    #   #<Reform::Contract::Errors:0x007fa739cc9100
    #    @base=#<#<Class:0x007fa73fbc92e0>:0x007fa73d3d8150 ...>,
    #    @messages={}>,
    #  @fields=
    #   {"discipline_class_id"=>517429222,
    #    "judge_id"=>301712148,
    #    "name"=>"Updated Competition",
    #    "description"=>"I am a new description.",
    #    "start_at"=>Wed, 01 Jul 2015 00:00:00 UTC +00:00,
    #    "end_at"=>Thu, 30 Jul 2015 23:59:59 UTC +00:00,
    #    "example_video"=>
    #     #<#<Class:0x007fa73fb4a7b0>:0x007fa73d3d3420
    #      @_changes={"video_file"=>true},
    #      @errors=
    #       #<Reform::Contract::Errors:0x007fa739cb1d70
    #        @base=#<#<Class:0x007fa73fb4a7b0>:0x007fa73d3d3420 ...>,
    #        @messages={}>,
    #      @fields={"video_file"=>""},
    #      @model=
    #       #<Video:0x007fa73d3d36f0
    #        id: 966191005,
    #        meta_info: {},
    #        created_at: Sun, 14 Jun 2015 09:00:32 UTC +00:00,
    #        updated_at: Sun, 14 Jun 2015 09:00:32 UTC +00:00,
    #        videoable_id: 802290399,
    #        videoable_type: "Competition",
    #        video_file_id: nil,
    #        video_file_filename: nil,
    #        video_file_content_type: nil>,
    #      @validation_context=nil>,
    #    "moves"=>
    #     [#<#<Class:0x007fa73fbd9050>:0x007fa73d3ca5a0
    #       @_changes={},
    #       @errors=
    #        #<Reform::Contract::Errors:0x007fa739cc8958
    #         @base=#<#<Class:0x007fa73fbd9050>:0x007fa73d3ca5a0 ...>,
    #         @messages={}>,
    #       @fields={"name"=>"Middle", "position"=>2, "id"=>908005748},
    #       @model=
    #        #<Move:0x007fa73d3cbab8
    #         id: 908005748,
    #         name: "Middle",
    #         position: 2,
    #         competition_id: 802290399,
    #         created_at: Sun, 14 Jun 2015 09:00:32 UTC +00:00,
    #         updated_at: Sun, 14 Jun 2015 09:00:32 UTC +00:00>,
    #       @validation_context=nil>,
    #      #<#<Class:0x007fa73fbd9050>:0x007fa73d3c9dd0
    #       @_changes={"name"=>true, "position"=>true, "id"=>true},
    #       @errors=
    #        #<Reform::Contract::Errors:0x007fa739cc3b38
    #         @base=#<#<Class:0x007fa73fbd9050>:0x007fa73d3c9dd0 ...>,
    #         @messages={}>,
    #       @fields={"name"=>"Middle Changed", "position"=>2, "id"=>"908005748"},
    #       @model=
    #        #<Move:0x007fa73d3cb8b0
    #         id: 908005746,
    #         name: "Last",
    #         position: 3,
    #         competition_id: 802290399,
    #         created_at: Sun, 14 Jun 2015 09:00:32 UTC +00:00,
    #         updated_at: Sun, 14 Jun 2015 09:00:32 UTC +00:00>,
    #       @validation_context=nil>,
    #      #<#<Class:0x007fa73fbd9050>:0x007fa739cc2f58
    #       @_changes={"name"=>true, "position"=>true, "id"=>true},
    #       @errors=
    #        #<Reform::Contract::Errors:0x007fa739cc97e0
    #         @base=#<#<Class:0x007fa73fbd9050>:0x007fa739cc2f58 ...>,
    #         @messages={}>,
    #       @fields={"name"=>"Last Changed", "position"=>3, "id"=>"908005746"},
    #       @model=
    #        #<Move:0x007fa739cc3188
    #         id: nil,
    #         name: nil,
    #         position: nil,
    #         competition_id: nil,
    #         created_at: nil,
    #         updated_at: nil>,
    #       @validation_context=nil>]},
    #  @model=
    #   #<Competition:0x007fa73d3d8380
    #    id: 802290399,
    #    discipline_class_id: 517429222,
    #    judge_id: 301712148,
    #    name: "Updated Competition",
    #    description: "I am a description.",
    #    pattern: nil,
    #    start_at: Fri, 12 Jun 2015 00:00:00 UTC +00:00,
    #    end_at: Fri, 19 Jun 2015 23:59:59 UTC +00:00,
    #    created_at: Sun, 14 Jun 2015 09:00:32 UTC +00:00,
    #    updated_at: Sun, 14 Jun 2015 09:00:32 UTC +00:00>,
    #  @validation_context=nil>
    
    pp model.moves if params[:competition]['moves'][0]['_destroy']
    # [#<Move:0x007fa73d3cbce8
    #   id: 908005747,
    #   name: "First",
    #   position: 1,
    #   competition_id: 802290399,
    #   created_at: Sun, 14 Jun 2015 09:00:32 UTC +00:00,
    #   updated_at: Sun, 14 Jun 2015 09:00:32 UTC +00:00>,
    #  #<Move:0x007fa73d3cbab8
    #   id: 908005748,
    #   name: "Middle",
    #   position: 2,
    #   competition_id: 802290399,
    #   created_at: Sun, 14 Jun 2015 09:00:32 UTC +00:00,
    #   updated_at: Sun, 14 Jun 2015 09:00:32 UTC +00:00>,
    #  #<Move:0x007fa73d3cb8b0
    #   id: 908005746,
    #   name: "Last",
    #   position: 3,
    #   competition_id: 802290399,
    #   created_at: Sun, 14 Jun 2015 09:00:32 UTC +00:00,
    #   updated_at: Sun, 14 Jun 2015 09:00:32 UTC +00:00>]
    
    f.save
    pp model.moves if params[:competition]['moves'][0]['_destroy']
    # [#<Move:0x007fa73d3cbab8
    #   id: 908005748,
    #   name: "Middle",
    #   position: 2,
    #   competition_id: 802290399,
    #   created_at: Sun, 14 Jun 2015 09:00:32 UTC +00:00,
    #   updated_at: Sun, 14 Jun 2015 09:00:32 UTC +00:00>,
    #  #<Move:0x007fa73d3cb8b0
    #   id: 908005746,
    #   name: "Middle Changed",
    #   position: 2,
    #   competition_id: 802290399,
    #   created_at: Sun, 14 Jun 2015 09:00:32 UTC +00:00,
    #   updated_at: Sun, 14 Jun 2015 09:00:32 UTC +00:00>,
    #  #<Move:0x007fa739cc3188
    #   id: 908005749,
    #   name: "Last Changed",
    #   position: 3,
    #   competition_id: 802290399,
    #   created_at: Sun, 14 Jun 2015 09:00:32 UTC +00:00,
    #   updated_at: Sun, 14 Jun 2015 09:00:32 UTC +00:00>]
  end
end
```

**Comment by apotonick** [comment]
Did we fix this with the new populator you use in Reform?

**Comment by kevinansfield** [comment]
Yep, all sorted with the latest reform/disposable versions, thanks.

My working collection code ended up like this:

```ruby
collection :moves,
  prepopulator: :prepopulate_moves!,
  populator: ->(fragment, collection, index, options) {
    collection.find { |m| m.id.to_s == fragment['id'] } || collection.insert(index, Move.new)
  },
  skip_if: :skip_move? do

  delegate :new_record?, to: :model

  property :name
  property :position

  validates :name, presence: true
end

private

def prepopulate_moves!(options)
  moves << model.moves.new if moves.size == 0
end

def skip_move?(fragment, options)
  fragment['name'].blank?
end
```
